File: client/src/modules/nexotype/schemas/omics/transcript.schemas.ts
import { z } from 'zod';

// ==========================================
// Full Schema (GET responses)
// ==========================================

export const TranscriptSchema = z.object({
  id: z.number(),
  gene_id: z.number().int(),
  ensembl_transcript_id: z.string().min(1).max(50),
  is_canonical: z.boolean(),
  created_at: z.string(),
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas (POST/PUT)
// ==========================================

export const CreateTranscriptSchema = z.object({
  gene_id: z.number().int(),
  ensembl_transcript_id: z.string().min(1).max(50),
  is_canonical: z.boolean().default(false),
});

export const UpdateTranscriptSchema = z.object({
  gene_id: z.number().int().optional(),
  ensembl_transcript_id: z.string().min(1).max(50).optional(),
  is_canonical: z.boolean().optional(),
});

// ==========================================
// Types
// ==========================================

export type Transcript = z.infer<typeof TranscriptSchema>;
export type CreateTranscript = z.infer<typeof CreateTranscriptSchema>;
export type UpdateTranscript = z.infer<typeof UpdateTranscriptSchema>;

// ==========================================
// Response Types
// ==========================================

export type TranscriptResponse = {
  success: boolean;
  data?: Transcript;
  error?: string;
};

export type TranscriptsResponse = {
  success: boolean;
  data?: Transcript[];
  count?: number;
  error?: string;
};

File: server/apps/nexotype/schemas/omics/transcript_schemas.py
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# ==========================================
# Request Schemas
# ==========================================

class TranscriptCreate(BaseModel):
    """Schema for creating a new transcript"""
    gene_id: int
    ensembl_transcript_id: str
    is_canonical: bool = False


class TranscriptUpdate(BaseModel):
    """Schema for updating a transcript"""
    gene_id: int = None
    ensembl_transcript_id: str = None
    is_canonical: bool = None


# ==========================================
# Response Schemas
# ==========================================

class TranscriptDetail(BaseModel):
    """Schema for transcript details"""
    id: int
    gene_id: int
    ensembl_transcript_id: str
    is_canonical: bool
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


class TranscriptListResponse(BaseModel):
    """Response schema for listing transcripts"""
    success: bool
    data: list[TranscriptDetail] | None = None
    count: int = 0
    error: str | None = None


class TranscriptResponse(BaseModel):
    """Response schema for single transcript operations"""
    success: bool
    data: TranscriptDetail | None = None
    error: str | None = None


class MessageResponse(BaseModel):
    """Simple message response"""
    success: bool
    message: str | None = None
    error: str | None = None
